Sources/Sentry/SentryOutOfMemoryTracker.m

```diff
@@ -0,0 +1,179 @@
+    if (![currentAppState.appVersion isEqualToString:previousAppState.appVersion]) {
+        return;
+    }
+
+    // The OS was upgraded
+    if (![currentAppState.osVersion isEqualToString:previousAppState.osVersion]) {
+        return;
+    }
+
+    // Restarting the app in development is a termination we can't catch and would falsely report
+    // OOMs.
+    if (previousAppState.isDebugging) {
+        return;
+    }
+
+    // The app was terminated normally
+    if (previousAppState.wasTerminated) {
+        return;
+    }
+
+    // The app crashed on the previous run. No OOM.
+    if (self.crashAdapter.crashedLastLaunch) {
+        return;
+    }
+
+    // Was the app in foreground/active ?
+    // If the app was in background we can't reliably tell if it was an OOM or not.
+    if (!previousAppState.isActive) {
+        return;
+    }
+
+    SentryEvent *event = [[SentryEvent alloc] initWithLevel:kSentryLevelFatal];
+    // Set to empty list so no breadcrumbs of the current scope are added
+    event.breadcrumbs = @[];
+
+    SentryException *exception =
+        [[SentryException alloc] initWithValue:SentryOutOfMemoryExceptionValue
+                                          type:SentryOutOfMemoryExceptionType];
+    SentryMechanism *mechanism =
+        [[SentryMechanism alloc] initWithType:SentryOutOfMemoryExceptionType];
+    mechanism.handled = @(NO);
+    exception.mechanism = mechanism;
+    event.exceptions = @[ exception ];
+
+    [SentrySDK captureCrashEvent:event];
+
+#else
+    [SentryLog logWithMessage:@"NO UIKit -> SentryOutOfMemoryTracker will not track OOM."
+                     andLevel:kSentryLevelInfo];
+    return;
+#endif
+}
+
+- (void)stop
+{
+#if SENTRY_HAS_UIKIT
+    [NSNotificationCenter.defaultCenter removeObserver:self];
+#endif
+}
+
+/**
+ * It is called when an App. is receiving events / It is in the foreground and when we receive a
+ * SentryHybridSdkDidBecomeActiveNotification.
+ */
+- (void)didBecomeActive
+{
+    // We don't know if the hybrid SDKs post the notification from a background thread, so we
+    // synchronize to be safe.
+    @synchronized(self) {
+        [self updateAppState:^(SentryAppState *appState) { appState.isActive = YES; }];
+    }
+}
+
+/**
+ * The app is about to lose focus / going to the background. This is only called when an app was
+ * receiving events / was is in the foreground.
+ */
+- (void)willResignActive
+{
+    [self updateAppState:^(SentryAppState *appState) { appState.isActive = NO; }];
+}
+
+- (void)willTerminate
+{
+    [self updateAppState:^(SentryAppState *appState) { appState.wasTerminated = YES; }];
+}
+
+- (void)updateAppState:(void (^)(SentryAppState *))block
+{
+    SentryFileManager *fileManager = [[[SentrySDK currentHub] getClient] fileManager];
+    SentryAppState *appState = [fileManager readAppState];
+    if (nil != appState) {
+        block(appState);
+        [fileManager storeAppState:appState];
```

Comment:
L: does that mean we do serialization and IO in the main thread for every received notification?

Reply:
Upsi, yes. I would change this to `h` then 🙈. I fix it.

Reply:
Oh my, yes it does. I think we must change this. I opened up a PR to be able to use a dispatch queue with a higher priority for updating the app state, https://github.com/getsentry/sentry-cocoa/pull/1003

Reply:
I fixed this now.